lib/hierarchical.h

```diff
@@ -465,6 +477,117 @@ class HierarchicalRadiosity {
         }
     }
 
+    /**
+     * Refine all links in all nodes.
+     * @return true if links have been refined.
+     */
+    bool refine_links() {
+        bool done = true;
+        for (auto& p : nodes_) {
+            if(!refine_links(p)) {
+                done = false;
+            }
+        }
+        return done;
+    }
+
+    /**
+     * Refine all links in node p.
+     * @return true if a link has been refined.
+     */
+    bool refine_links(Quadnode& p) {
+        bool done = true;
+
+        // We create a copy of gathering links because refine_link might remove
+        // elements.
+        // TODO: Avoid the copy
+        std::vector<Linknode> links = p.gathering_from;
+        for (auto& link: links) {
+            if (!refine_link(p, link)) {
+                done = false;
+            }
+        }
+
+        // Recursive for all child nodes.
+        // TODO: Use stack instead.
+        // TODO: BUG children will have refined links already. BAD.
+        if (!p.is_leaf()) {
+            for (auto& child: p.children) {
+                if(!refine_links(*child.get())) {
+                    done = false;
+                }
+            }
+        }
+        return done;
+    }
+
+    /**
+     * Refine link of receiver node p.
+     *
+     * @param q receiver node
+     * @param link link between shooter and receiver node
+     * @return true if a link has been refined.
+     */
+    bool refine_link(Quadnode& q, Linknode& link_node) {
+        bool no_subdivision = true;
+
+        // Shooter node p
+        Quadnode& p = *link_node.q;
+
+        const Triangle& tri_q = get_triangle(q);
+        const Triangle& tri_p = get_triangle(p);
+
+        auto oracle = p.rad_shoot * tri_p.area() * link_node.form_factor;
+        //std::cerr << oracle << std::endl;
+        if (oracle.r > BF_eps_ || oracle.g > BF_eps_ || oracle.b > BF_eps_) {
+            std::cerr << "Refine link.." << std::endl;
+            no_subdivision = false;
+
+            float F_qp = link_node.form_factor;
+            float F_pq = F_qp * tri_q.area() / tri_p.area();
+
+            // Remove p from q's gather nodes. It's safe because we are
+            // iterating over it's copy.
+            // TODO: Make gather_from an unordered set for fast delete.
+            std::cerr << "Remove link." << std::endl;
+            for (auto it = q.gathering_from.begin(); it != q.gathering_from.end();) {
+                if (it->q->tri_id == p.tri_id) {
+                    it = q.gathering_from.erase(it);
```

Comment:
Da wir in `bool refine_links(Quadnode& p)` über eine Kopier der Links iterieren, ist das hier okay. Wäre aber schöne ohne Kopie auszukommen.